Add metadata to BasicGameSaveGameInfo(Widget)

ZashIn opened the pull request:
Adds optional metadata to the `BasicGameSaveGameInfoWidget`, similar to TES / Gamebryo games:
```py
BasicGameSaveGameInfo()  # Adds "File Date" by default
BasicGameSaveGameInfo(
    get_preview = lambda p: Path(p, "screenshot.png"), # Save preview
    get_metadata = lambda p, save: {"Name": save.getName(), ...}
)
```

Example from Cyberpunk 2077, included in #105:
![image](https://github.com/ModOrganizer2/modorganizer-basic_games/assets/10155187/6dad33e8-4eeb-4dce-8784-d4f72768e590)

Adds a default file date to all BasicGames, taken from the save game files (`GameSaveExtension` etc.)

Holt59 commented on basic_features/basic_save_game_info.py:
```diff
@@ -33,70 +51,133 @@ def allFiles(self) -> list[str]:
+    def __init__(
+        self,
+        parent: QWidget,
+        get_preview: Callable[[Path], Path | None] | None = None,
```
You have a double `None` check which is redundant, you can simply default to `lambda p: None` and remove the possibility of having `None`.

This also applies to `get_preview`.

ZashIn replied:
Replaced the double `None` check and added all supported types to `get_preview`
https://github.com/ModOrganizer2/modorganizer-basic_games/blob/dd95e5a6aee4ffe27d1c6ac98ff03105f68fb24d/basic_features/basic_save_game_info.py#L59-L64

Holt59 commented:
Sorry for the delay, could you try rebasing/merging against latest master so that these change can be merged? If you don't have time, I can have a look at it myself.

ZashIn commented:
> Sorry for the delay, could you try rebasing/merging against latest master so that these change can be merged? If you don't have time, I can have a look at it myself.

I can probably get to it tomorrow, have to check python 11 changes (my local poetry will be conflicting, too).
If I remember correctly, I made this as backwards compatible as possible, but there is at least one game, which implements `BasicGameSaveGameInfoWidget`, that should be tested.

From a quick search that would still be only: `Black & White 2`.

ZashIn commented:
Rebase done. No problems besides the ugly `None` vs default param checks above.

Holt59 commented:
Thanks for the update! 

@Silarn Added save metadata preview to Blade & Sorcery recently, could you see if these could be moved to your new generic metadata system?

Silarn commented:
Obviously this would need to be merged in first but yeah, I think you can just migrate the BaSSaveGameWidget code to a new 'bas_parse_metadata' function passed into the updated BasicSaveGameInfo.

ZashIn commented:
> Obviously this would need to be merged in first but yeah, I think you can just migrate the BaSSaveGameWidget code to a new 'bas_parse_metadata' function passed into the updated BasicSaveGameInfo.

I added an option to set the (max) widget width and refactored blade & sorcery, should work.
Cleaned up Black & White 2, since it was not actually using its minimal custom widget.

Want me to cleanup the history before merge?

Holt59 commented on basic_features/basic_save_game_info.py:
```diff
@@ -31,81 +49,166 @@ def allFiles(self) -> list[str]:
         super().__init__()
         self._get_preview = get_preview
+        self._get_metadata = get_metadata
+        self._maximum_widget_size = 0
```
Any reason to not pass this to the constructor and use `with_maximum_widget_size()`?

Holt59 commented on basic_features/basic_save_game_info.py:
```diff
@@ -31,81 +49,166 @@ def allFiles(self) -> list[str]:
     def __init__(
         self,
         parent: QWidget | None,
         get_preview: Callable[
```
Maybe change this to also get the savegame to be consistent between `get_preview` and `get_metadata`.

ZashIn commented:
Ok, widget size, default parameter handling and an easier option to set a custom widget added:
`BasicGameSaveGameInfo.with_widget(ISaveGameInfoWidget)`

Widget tested with Cyberpunk.
A test for Blade & Sorcery would be good, with only the metadata.

Silarn commented:
Unfortunately I was having a requester on Discord test my initial changes. But I might be able to get a hold of him.

ZashIn commented:
Commits cleaned up. Ready to merge (depending on tests)

ZashIn commented:
I removed Cyberpunk `BasicGameSaveGameInfo` from this PR, to be included in #105.

ZashIn commented:
Rebased on master, with PR #122 merged.
